Rumble to Intake

Yumnah2 opened the pull request.

Yumnah2 committed:
Added rumble to drive controller

driver controller rumbles when feeder sensor senses note and intake stops

Yumnah2 committed:
Merge branch 'main' of https://github.com/Yumnah2/Crescendo2024

Yumnah2 committed:
merge conflicts

:C

TAKBS2412 commented on src/main/java/frc/team2412/robot/Controls.java:
```diff
@@ -121,11 +124,15 @@ private void bindDrivebaseControls() {
 		driveIntakeReverseButton.onTrue(new AllReverseCommand(s.intakeSubsystem));
 		driveIntakeRejectButton.onTrue(new IntakeRejectCommand(s.intakeSubsystem));
-		codriveIntakeInButton.onTrue(new AllInCommand(s.intakeSubsystem));
+		codriveIntakeInButton.onTrue(new AllInCommand(s.intakeSubsystem, this));
```
Curious why this is different from the drive intake in button?

TAKBS2412 commented on src/main/java/frc/team2412/robot/util/AutoLogic.java:
```diff
@@ -65,7 +65,7 @@ public void registerCommands() {
 						.andThen(new FeederInCommand(s.intakeSubsystem)));
 
-		NamedCommands.registerCommand("Intake", new AllInCommand(s.intakeSubsystem));
+		NamedCommands.registerCommand("Intake", new AllInCommand(s.intakeSubsystem, controls));
```
Note that since this is in auto we don't need to rumble the drive controller... probably okay to leave in though.

jamesdooley4 replied:
`AllInCommand` could be updated to tolerate `null` being passed for `controls` and skip the controller vibration in that case. That'd avoid the controllers vibrating during auto and potentially falling off the drive board.

jamesdooley4 commented on src/main/java/frc/team2412/robot/commands/intake/AllInCommand.java:
```diff
@@ -30,6 +33,7 @@ public void end(boolean interrupted) {
 		intakeSubsystem.intakeReject();
 		intakeSubsystem.indexStop();
 		intakeSubsystem.feederStop();
+		controls.vibrateDriveController(1.0);
```
Since the command can be interrupted for reasons other than `isFinished()` being signaled (could be aborted if scheduled in parallel with a wait command, a different button is pressed that starts a different command using the intake subsystem, etc.), it'd be best to check and only rumble the controller if `intakeSubsystem.feederSensorHasNote()` is true.

jamesdooley4 replied:
Or, more simply, only vibrate the controller if the `interrupted` parameter is `false` (meaning the command is ending because it finished and wasn't interrupted).

jamesdooley4 commented on src/main/java/frc/team2412/robot/Controls.java:
```diff
@@ -121,11 +124,15 @@ private void bindDrivebaseControls() {
 		// CommandScheduler.getInstance()
 		// 		.setDefaultCommand(s.intakeSubsystem, new IntakeStopCommand(s.intakeSubsystem));
-		driveIntakeInButton.onTrue(new AllInCommand(s.intakeSubsystem));
+		driveIntakeInButton.onTrue(
+				Commands.sequence(
+						new AllInCommand(s.intakeSubsystem, this),
+						new WaitCommand(0.5),
+						new StopRumbleCommand(this)));
```
I'm pretty sure the controller would get stuck vibrating if the driver or codriver pressed any of the other intake buttons right after a note is detected & the controller starts rumbling. That would interrupt this sequence during the WaitCommand so the StopRumbleCommand would never run. One way to fix that is to wrap AllInCommand and WaitCommand in a new command class derived from SequentialCommandGroup and have the `end(..)` method call controls.vibrateDriveController(0.0).

Yumnah2 committed:
Changed command structure for rumble

addressing comments by adding AllInRumbleCommand

jamesdooley4 commented on src/main/java/frc/team2412/robot/commands/intake/AllInRumbleCommand.java:
```diff
@@ -0,0 +1,23 @@
+	public AllInRumbleCommand(IntakeSubsystem intakeSubsystem, Controls controls) {
+		this.intakeSubsystem = intakeSubsystem;
+		this.controls = controls;
+		addRequirements(intakeSubsystem);
```
This implicitly happens in the `addCommands` call below (see https://github.com/wpilibsuite/allwpilib/blob/f1a1ffd7fc96649df2293e6337e02f25b557682c/wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/SequentialCommandGroup.java#L51 ) so should be removed here

Yumnah2 committed:
Stop intake rumble in auto

Yumnah2 committed:
deleting addRequirements in AllInRumbleCommand

Yumnah2 committed:
Added AllInNoSensorCommand

intaking without sensors for auto usage

TAKBS2412 commented on src/main/java/frc/team2412/robot/commands/intake/AllInRumbleCommand.java:
```diff
@@ -0,0 +1,22 @@
+		addCommands(new AllInCommand(intakeSubsystem, controls), new WaitCommand(0.5));
+	}
+
+	public void end() {
```
Note that we will want to call the superclass's `end()` method to make sure the command group responds to cancellations properly.

TAKBS2412 commented on src/main/java/frc/team2412/robot/commands/intake/AllInCommand.java:
```diff
@@ -30,6 +33,10 @@ public void end(boolean interrupted) {
 		intakeSubsystem.indexStop();
 		intakeSubsystem.feederStop();
+
+		if (controls != null && interrupted == false) {
```
(Nit: you can use `!interrupted` instead of directly comparing against `false`.)

Yumnah2 committed:
rewrote interrupted in AllInCommand

iamawesomecat commented:
We should prob get this merged today

Yumnah2 committed:
Merge branch 'main' of https://github.com/Yumnah2/Crescendo2024

Yumnah2 committed:
stop rumble command

ignore this actually

Yumnah2 committed:
Rumble and intake sensor logic

All rollers reject after intake sensor logic and we included rumble through different stages

Yumnah2 committed:
removed rumble during auto

no rumbling during auto

jamesdooley4 commented on src/main/java/frc/team2412/robot/commands/intake/AllInCommand.java:
```diff
@@ -20,16 +25,66 @@ public void initialize() {
+			}
+		}
+
+		if (intakeSubsystem.intakeBackSeesNote()) {
```
What should happen if more than one side sees a note?

Yumnah2 replied:
We talked to Dom and he said that one sensor would probably sense a note before another one, and therefore the second sensor and intake roller would start rejecting the second note. But we should probably test it on the robot to be sure

jamesdooley4 commented on src/main/java/frc/team2412/robot/commands/intake/AllInCommand.java:
```diff
@@ -20,16 +25,66 @@ public void initialize() {
 	@Override
 	public void end(boolean interrupted) {
-		// intakeSubsystem.intakeReject();
+		intakeSubsystem.intakeReject();
```
Since stop is called on the next line, the intake motors won't have time to reverse, and will just stop. Recommend deleting the call to intakeReject.

TAKBS2412 replied:
I think this is intake reject followed by index stop, so the intake motors will keep rejecting.

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -140,6 +147,11 @@ public IntakeSubsystem() {
+		intakeFrontSensor = intakeMotorFront.getForwardLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
+		intakeBackSensor = intakeMotorBack.getForwardLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
+		intakeLeftSensor = intakeMotorLeft.getForwardLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
+		intakeRightSensor = intakeMotorRight.getForwardLimitSwitch(SparkLimitSwitch.Type.kNormallyOpen);
```
Also need to update the motor configuration code so the controller knows the limit switch is plugged in (the same wire can also be used for an encoder), and to not stop the motor when the switch is signaled

Yumnah2 replied:
We updated the configuration but not sure if it'll impact the motor

Yumnah2 replied:
enableLimitSwitch() is set to false so it shouldn't impact the motors

Yumnah2 committed:
added intake Reject override, rumble debugging

rumble debugging- rumbles during auto (will fix that later), increased rumble time, rumble to both controllers

TAKBS2412 commented on src/main/java/frc/team2412/robot/Controls.java:
```diff
@@ -122,11 +122,12 @@ private void bindDrivebaseControls() {
 		// 		.setDefaultCommand(s.intakeSubsystem, new IntakeStopCommand(s.intakeSubsystem));
-		driveIntakeInButton.onTrue(new AllInCommand(s.intakeSubsystem));
-		// driveIntakeStopButton.onTrue(new AllStopCommand(s.intakeSubsystem));
+		driveIntakeInButton.onTrue(new RumbleCommand(this));
```
Should this be `AllInCommand`?

Yumnah2 replied:
while testing we changed it to see if rumble worked on the controllers so we'll change it back to AllInCommand soon

jamesdooley4 replied:
Need to change it back

TAKBS2412 commented on src/main/java/frc/team2412/robot/Subsystems.java:
```diff
@@ -14,8 +14,8 @@ public static class SubsystemConstants {
 		public static final boolean APRILTAGS_ENABLED = true;
-		public static final boolean LIMELIGHT_ENABLED = false;
-		public static final boolean CLIMB_ENABLED = false;
+		public static final boolean LIMELIGHT_ENABLED = true;
+		public static final boolean CLIMB_ENABLED = true;
```
These should be reverted to `false`.

Yumnah2 replied:
fixed, thanks

TAKBS2412 commented on src/main/java/frc/team2412/robot/commands/intake/AllInCommand.java:
```diff
@@ -20,16 +25,98 @@ public void initialize() {
+			}
+
+			if (controls != null) {
+				Commands.race(new RumbleCommand(controls), new WaitCommand(1)).schedule();
```
FYI that you can use [`withTimeout()`](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/Command.html#withTimeout(double)) (not required, but useful tip).

TAKBS2412 commented on src/main/java/frc/team2412/robot/commands/intake/AllInCommand.java:
```diff
@@ -20,16 +25,98 @@ public void initialize() {
+			}
+
+			if (controls != null) {
+				Commands.race(new RumbleCommand(controls), new WaitCommand(3)).schedule();
```
Is there a reason why the rumble here is for 3 seconds, but for back, left, and right it's 2 seconds, and only 1 second for the front? (maybe a method of signaling to the drive team?)

Yumnah2 replied:
Sorry I forgot to change the intake front. The intake stops/rejects should be 2 seconds, index stopping/rejecting all of the intake should be 3, and end() should be 3 as well.

TAKBS2412 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -47,8 +48,13 @@ public class IntakeSubsystem extends SubsystemBase {
 	private final DigitalInput indexSensor;
 	private final DigitalInput feederSensor;
 
+	private final SparkLimitSwitch intakeFrontSensor;
+	private final SparkLimitSwitch intakeBackSensor;
+	private final SparkLimitSwitch intakeLeftSensor;
+	private final SparkLimitSwitch intakeRightSensor;
```
We should probably log these to help debug any issues in the future.

Yumnah2 replied:
added them to shuffleboard

TAKBS2412 commented on src/main/java/frc/team2412/robot/commands/intake/AllInCommand.java:
```diff
@@ -20,16 +25,98 @@ public void initialize() {
+			}
+
+			if (controls != null) {
+				Commands.race(new RumbleCommand(controls), new WaitCommand(2)).schedule();
```
Note that this will cause a new rumble command to be scheduled every ~20 ms until the note stops triggering the sensor, which would be a bit excessive (and might lead to a crash depending on the maximum number of commands).

Yumnah2 replied:
This should be fixed now

bobtseattle commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -140,11 +155,17 @@ public IntakeSubsystem() {
 		ShuffleboardTab shuffleboardTab = Shuffleboard.getTab("Intake");
 		shuffleboardTab.addBoolean("Index Sensor - ", this::indexSensorHasNote).withSize(1, 1);
 		shuffleboardTab.addBoolean("Feeder Sensor - ", this::feederSensorHasNote).withSize(1, 1);
+		shuffleboardTab.addBoolean("Intake Sensor - ", this::intakeBackSeesNote).withSize(1, 1);
```
Should this be back or if any intake sensors are triggered?

Yumnah2 replied:
Currently there are all of the sensors except the back one

Yumnah2 committed:
addressed github comments

added intake sensors to shuffleboard, removed intake back sensor, changed motor configuration and stopped continuous rumbling

TAKBS2412 reviewed (dismissed):
Approving to have a baseline before robot testing.

Note that the rumbled variables like `rumbledIntakeFront` aren't reset back to `false` once the command ends, so the controller will only rumble the first time the robot intakes a note on a given side. That can be fixed in a later PR though.

Yumnah2 committed:
reset rumbled variables to false after intaking

bobtseattle reviewed (dismissed):
LGTM

jamesdooley4 commented on src/main/java/frc/team2412/robot/Controls.java:
```diff
@@ -176,8 +177,10 @@ private void bindLauncherControls() {
-		if (!DriverStation.isAutonomous()) {
-			driveController.getHID().setRumble(RumbleType.kBothRumble, vibration);
-		}
+		// if (!DriverStation.isAutonomous()) {
+
+		// }
```
Should delete commented-out code

jamesdooley4 replied:
Or put it back so the controllers don't rumble in autonomous

Yumnah2 committed:
undid the rumble debugging code

bobtseattle reviewed (dismissed):
Let's get this in so we can merge with ingest disable code.

jamesdooley4 committed:
Merge remote-tracking branch 'origin/main' into pr/71

jamesdooley4 committed:
Fix conflicts with main

jamesdooley4 commented:
@Yumnah2 - I pushed commits to resolve the merge conflicts & autofix the spotless syntax updates
